### DISCUSSION TITLE
Inconsistent behavior in overriding keymaps for nvim-cmp

### DISCUSSION THREAD

**Discussion by Trildar** [LazyVim/LazyVim#1433] [Q&A] [answered]
When I try to change the `<C-n>` and `<C-p>` keymaps to use `cmp.SelectBehavior.Select` instead of `Insert`, I get varying results on every start of Neovim. Sometimes it is `Select` as expected, sometimes `Insert`, and strangest of all sometimes one key will have `Insert` behavior but the other `Select`; I think usually I notice `<C-n>` as being the one with `Insert` behavior.

Anyone have any idea if maybe I'm not doing this right, or there's an issue in lazy.nvim, or if it could actually be something to do with nvim-cmp?

Config below for reference.

```Lua
return {
  {
    "hrsh7th/nvim-cmp",
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.mapping = vim.tbl_deep_extend("force", opts.mapping, {
        -- Override selection behavior to not immediately insert
        ["<C-n>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Select }),
        ["<C-p>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Select }),
        ["<C-l>"] = cmp.mapping.complete(),
      })
    end,
  },
}
```

**Comment by dpetka2001** [comment, marked as answer]
It seems that the function `cmp.mapping.preset.insert()` that is used to define the keys already does some merging of its own, so it seems to me like there's some kind of race condition between `cmp`'s internal merge function and `vim.tbl_deep_extend()` (I might be wrong about this). I, like you, couldn't consistently observe the desired behavior from your keymaps. Some times it would behave like `Insert` and some times like `Select`. The best solution I found to this is take Lazyvim's default config and change what you want. So, in your spec you should do something like
```lua
{
    "hrsh7th/nvim-cmp",
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.mapping = cmp.mapping.preset.insert({
        ["<C-n>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
        ["<C-p>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Insert }),
        ["<C-b>"] = cmp.mapping.scroll_docs(-4),
        ["<C-f>"] = cmp.mapping.scroll_docs(4),
        ["<C-Space>"] = cmp.mapping.complete(),
        ["<C-e>"] = cmp.mapping.abort(),
        ["<CR>"] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
        ["<S-CR>"] = cmp.mapping.confirm({
          behavior = cmp.ConfirmBehavior.Replace,
          select = true,
        }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
      }),
    end,
  },
```
and change `<C-n>` and `<C-p` to `Select` like you want. Maybe there's a better solution, but I fail to think of it.

**Reply by Trildar** [reply]
Thanks. That seems to work. I wonder why though. I would have thought the user `opts` function would just run entirely after the one from LazyVim. I also tried assigning to `opts.mapping["<C-n>"]` (and \<C-p>) instead and got the same inconsistent behavior.
